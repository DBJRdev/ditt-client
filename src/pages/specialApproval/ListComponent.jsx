import Immutable from 'immutable';
import ImmutablePropTypes from 'react-immutable-proptypes';
import PropTypes from 'prop-types';
import React from 'react';
import jwt from 'jsonwebtoken';
import { Table } from 'react-ui';
import Layout from '../../components/Layout';
import {
  BUSINESS_TRIP_WORK_LOG,
  HOME_OFFICE_WORK_LOG,
  TIME_OFF_WORK_LOG,
} from '../../resources/workMonth';
import {
  toDayMonthYearFormat,
} from '../../services/dateTimeService';

class ListComponent extends React.Component {
  componentDidMount() {
    if (this.props.token) {
      const decodedToken = jwt.decode(this.props.token);

      if (decodedToken) {
        this.props.fetchSpecialApprovalList(decodedToken.uid);
      }
    }
  }

  getFilteredSpecialApprovals() {
    let specialApprovalList = Immutable.List();

    specialApprovalList = specialApprovalList.concat((
      this.props.specialApprovalList.get('businessTripWorkLogs').map((
        workLog => workLog
          .set('id', `BT-${workLog.get('id')}`)
          .set('type', BUSINESS_TRIP_WORK_LOG)
      ))
    ));
    specialApprovalList = specialApprovalList.concat((
      this.props.specialApprovalList.get('homeOfficeWorkLogs').map((
        workLog => workLog
          .set('id', `HO-${workLog.get('id')}`)
          .set('type', HOME_OFFICE_WORK_LOG)
      ))
    ));
    specialApprovalList = specialApprovalList.concat((
      this.props.specialApprovalList.get('timeOffWorkLogs').map((
        workLog => workLog
          .set('id', `TO-${workLog.get('id')}`)
          .set('type', TIME_OFF_WORK_LOG)
      ))
    ));
    specialApprovalList = specialApprovalList.sortBy(workLog => -workLog.get('date'));

    return specialApprovalList;
  }

  render() {
    const specialApprovals = this.getFilteredSpecialApprovals();

    return (
      <Layout title="Special approvals" loading={this.props.isFetching}>
        {specialApprovals.count() > 0 ? (
          <Table
            columns={[
              {
                format: row => `${row.workMonth.user.firstName} ${row.workMonth.user.lastName}`,
                label: 'Name',
                name: 'lastName',
              },
              {
                format: row => toDayMonthYearFormat(row.date),
                label: 'Date',
                name: 'date',
              },
              {
                format: (row) => {
                  switch (row.type) {
                    case BUSINESS_TRIP_WORK_LOG:
                      return 'Business trip';
                    case HOME_OFFICE_WORK_LOG:
                      return 'Home office';
                    case TIME_OFF_WORK_LOG:
                      return 'Time off';
                    default:
                      return '-';
                  }
                },
                label: 'Type',
                name: 'type',
              },
              {
                format: () => (
                  <div>
                    <span>Approved</span>
                    <span>Rejected</span>
                  </div>
                ),
                label: 'Actions',
                name: 'actions',
              },
            ]}
            rows={specialApprovals.toJS()}
          />
        ) : (
          <div>
            You do not seem to have any pending special approvals.
          </div>
        )}
      </Layout>
    );
  }
}

ListComponent.propTypes = {
  fetchSpecialApprovalList: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  specialApprovalList: ImmutablePropTypes.mapContains({
    businessTripWorkLogs: ImmutablePropTypes.listOf(ImmutablePropTypes.mapContains({
      date: PropTypes.shape.isRequired,
      id: PropTypes.number.isRequired,
      workMonth: ImmutablePropTypes.mapContains({
        user: ImmutablePropTypes.mapContains({
          firstName: PropTypes.string.isRequired,
          lastName: PropTypes.string.isRequired,
        }).isRequired,
      }).isRequired,
    })).isRequired,
    homeOfficeWorkLogs: ImmutablePropTypes.listOf(ImmutablePropTypes.mapContains({
      date: PropTypes.shape.isRequired,
      id: PropTypes.number.isRequired,
      workMonth: ImmutablePropTypes.mapContains({
        user: ImmutablePropTypes.mapContains({
          firstName: PropTypes.string.isRequired,
          lastName: PropTypes.string.isRequired,
        }).isRequired,
      }).isRequired,
    })).isRequired,
    timeOffWorkLogs: ImmutablePropTypes.listOf(ImmutablePropTypes.mapContains({
      date: PropTypes.shape.isRequired,
      id: PropTypes.number.isRequired,
      workMonth: ImmutablePropTypes.mapContains({
        user: ImmutablePropTypes.mapContains({
          firstName: PropTypes.string.isRequired,
          lastName: PropTypes.string.isRequired,
        }).isRequired,
      }).isRequired,
    })).isRequired,
  }).isRequired,
  token: PropTypes.string.isRequired,
};

export default ListComponent;
