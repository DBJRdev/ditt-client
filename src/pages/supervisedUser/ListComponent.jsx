import ImmutablePropTypes from 'react-immutable-proptypes';
import PropTypes from 'prop-types';
import React from 'react';
import jwt from 'jsonwebtoken';
import { Link } from 'react-router-dom';
import { withNamespaces } from 'react-i18next';
import { Table } from 'react-ui';
import Layout from '../../components/Layout';
import routes from '../../routes';
import {
  STATUS_APPROVED,
  STATUS_OPENED,
  STATUS_WAITING_FOR_APPROVAL,
} from '../../resources/workMonth';
import {
  createDate,
  toMonthYearFormat,
} from '../../services/dateTimeService';
import styles from './supervisedUser.scss';

class ListComponent extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      tableSortColumn: 'lastName',
      tableSortDirection: 'asc',
    };
  }

  componentDidMount() {
    if (this.props.token) {
      const decodedToken = jwt.decode(this.props.token);

      if (decodedToken) {
        this.props.fetchSupervisedUserList(
          decodedToken.uid,
          {
            order: {
              column: this.state.tableSortColumn,
              direction: this.state.tableSortDirection,
            },
          }
        );
      }
    }
  }

  render() {
    const { t } = this.props;

    let uid = null;

    if (this.props.token) {
      const decodedToken = jwt.decode(this.props.token);

      if (decodedToken) {
        // eslint-disable-next-line prefer-destructuring
        uid = decodedToken.uid;
      }
    }

    const lighterRow = (row) => {
      if (!row.supervisor) {
        return '';
      }

      return row.supervisor.id !== uid ? styles.lighterRow : '';
    };

    return (
      <Layout title={t('supervisedUser:title.supervisedUsers')} loading={this.props.isFetching}>
        {this.props.supervisedUserList.count() > 0 ? (
          <Table
            columns={[
              {
                format: row => (
                  <span className={lighterRow(row)}>
                    {`${row.firstName} ${row.lastName}`}
                  </span>
                ),
                isSortable: true,
                label: t('user:element.name'),
                name: 'lastName',
              },
              {
                format: (row) => {
                  const waitingForApproval = row.workMonths.filter(workMonth => (
                    workMonth.status === STATUS_WAITING_FOR_APPROVAL
                  ));

                  if (!waitingForApproval.length) {
                    return (
                      <span className={lighterRow(row)}>
                        {t('general:action.no')}
                      </span>
                    );
                  }

                  const waitingForApprovalLinks = waitingForApproval.map(workMonth => (
                    /* eslint-disable-next-line jsx-a11y/anchor-is-valid */
                    <Link
                      className={styles.waitingForApprovalLink}
                      key={workMonth.id}
                      to={
                        routes.supervisedUserWorkLogWithDate
                          .replace(':id', row.id)
                          .replace(':year', workMonth.year)
                          .replace(':month', workMonth.month)
                      }
                    >
                      {toMonthYearFormat(createDate(workMonth.year, workMonth.month - 1, 1))}
                    </Link>
                  ));

                  return (
                    <div className={lighterRow(row)}>
                      {t('general:action.yes')}
                      {' | '}
                      {waitingForApprovalLinks}
                    </div>
                  );
                },
                label: t('supervisedUser:element.needApproval'),
                name: 'needApproval',
              },
              {
                format: row => (
                  <span className={lighterRow(row)}>
                    {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
                    <Link to={routes.supervisedUserWorkLog.replace(':id', row.id)}>
                      {t('supervisedUser:element.showWorkLog')}
                    </Link>
                  </span>
                ),
                label: t('supervisedUser:element.showWorkLog'),
                name: 'showWorkLog',
              },
            ]}
            rows={this.props.supervisedUserList.toJS()}
            sort={{
              changeHandler: (column, direction) => {
                if (this.props.token) {
                  const decodedToken = jwt.decode(this.props.token);

                  if (decodedToken) {
                    const orderDirection = direction === 'asc' ? 'desc' : 'asc';

                    this.props.fetchSupervisedUserList(
                      decodedToken.uid,
                      {
                        order: {
                          column,
                          direction: orderDirection,
                        },
                      }
                    ).then(() => {
                      this.setState({
                        tableSortColumn: column,
                        tableSortDirection: orderDirection,
                      });
                    });
                  }
                }
              },
              column: this.state.tableSortColumn,
              direction: this.state.tableSortDirection,
            }}
          />
        ) : (
          <div>
            {t('supervisedUser:text.emptyList')}
          </div>
        )}
      </Layout>
    );
  }
}

ListComponent.propTypes = {
  fetchSupervisedUserList: PropTypes.func.isRequired,
  isFetching: PropTypes.bool.isRequired,
  supervisedUserList: ImmutablePropTypes.listOf(ImmutablePropTypes.mapContains({
    firstName: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    lastName: PropTypes.string.isRequired,
    supervisor: ImmutablePropTypes.mapContains({
      firstName: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
      lastName: PropTypes.string.isRequired,
    }),
    workMonths: ImmutablePropTypes.listOf(ImmutablePropTypes.mapContains({
      status: PropTypes.oneOf([
        STATUS_APPROVED,
        STATUS_OPENED,
        STATUS_WAITING_FOR_APPROVAL,
      ]).isRequired,
    })).isRequired,
  })).isRequired,
  t: PropTypes.func.isRequired,
  token: PropTypes.string.isRequired,
};

export default withNamespaces()(ListComponent);
