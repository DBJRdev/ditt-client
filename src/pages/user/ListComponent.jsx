import moment from 'moment-timezone';
import jwt from 'jsonwebtoken';
import ImmutablePropTypes from 'react-immutable-proptypes';
import PropTypes from 'prop-types';
import React from 'react';
import {
  Button,
  Table,
} from 'react-ui';
import { Link } from 'react-router-dom';
import routes from '../../routes';
import Layout from '../../components/Layout';
import { ROLE_SUPER_ADMIN } from '../../resources/user';
import styles from './user.scss';

class ListComponent extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      tableSortColumn: 'lastName',
      tableSortDirection: 'asc',
    };
  }

  componentDidMount() {
    this.props.fetchUserList({
      order: {
        column: this.state.tableSortColumn,
        direction: this.state.tableSortDirection,
      },
    });
  }

  render() {
    const year = moment().year();
    const columns = [
      {
        format: row => (
          /* eslint-disable-next-line jsx-a11y/anchor-is-valid */
          <Link to={routes.editUser.replace(':id', row.id)}>
            {row.firstName} {row.lastName}
          </Link>
        ),
        isSortable: true,
        label: 'Name',
        name: 'lastName',
      },
      {
        format: (row) => {
          if (row.supervisor) {
            return `${row.supervisor.firstName} ${row.supervisor.lastName}`;
          }

          return '-';
        },
        isSortable: true,
        label: 'Supervisor',
        name: 'supervisor.lastName',
      },
      {
        format: (row) => {
          const userYearStats = row.yearStats.filter(stats => stats.year === year)[0];

          return `${Math.round(userYearStats.workedHours * 10) / 10}/${userYearStats.requiredHours}`;
        },
        isSortable: false,
        label: 'Worked / Required hours',
        name: 'requiredWorkedHours',
      },
      {
        format: row =>
          `${row.yearStats.filter(stats => stats.year === year)[0].vacationDaysUsed}/${row.vacationDays}`,
        isSortable: false,
        label: 'Vacation days used',
        name: 'vacationDaysUsed',
      },
    ];

    if (this.props.token) {
      const decodedToken = jwt.decode(this.props.token);

      if (decodedToken && decodedToken.roles.some(role => ROLE_SUPER_ADMIN === role)) {
        columns.push({
          format: row => (
            /* eslint-disable-next-line jsx-a11y/anchor-is-valid */
            <Link to={routes.supervisedUserWorkLog.replace(':id', row.id)}>
              Show work log
            </Link>
          ),
          label: 'Show work log',
          name: 'showWorkLog',
        });
      }
    }

    return (
      <Layout title="Users" loading={this.props.isFetching}>
        <div className={styles.actions}>
          <Button
            clickHandler={() => this.props.history.push(routes.addUser)}
            label="Add user"
            priority="primary"
          />
        </div>
        <Table
          columns={columns}
          rows={this.props.userList.toJS()}
          sort={{
            changeHandler: (column, direction) => {
              const orderDirection = direction === 'asc' ? 'desc' : 'asc';

              this.props.fetchUserList({
                order: {
                  column,
                  direction: orderDirection,
                },
              }).then(() => {
                this.setState({
                  tableSortColumn: column,
                  tableSortDirection: orderDirection,
                });
              });
            },
            column: this.state.tableSortColumn,
            direction: this.state.tableSortDirection,
          }}
        />
      </Layout>
    );
  }
}

ListComponent.propTypes = {
  fetchUserList: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  isFetching: PropTypes.bool.isRequired,
  token: PropTypes.string.isRequired,
  userList: ImmutablePropTypes.listOf(ImmutablePropTypes.mapContains({
    firstName: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    lastName: PropTypes.string.isRequired,
    supervisor: ImmutablePropTypes.mapContains({
      firstName: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
      lastName: PropTypes.string.isRequired,
    }),
  })).isRequired,
};

export default ListComponent;
