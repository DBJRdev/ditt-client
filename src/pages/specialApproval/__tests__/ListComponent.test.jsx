import Immutable from 'immutable';
import React from 'react';
import { shallow } from 'enzyme';
import ListComponent from '../ListComponent';

describe('rendering', () => {
  it('renders correctly', () => {
    const props = {
      fetchSpecialApprovalList: () => {},
      isFetching: false,
      specialApprovalList: Immutable.fromJS({
        businessTripWorkLogs: [],
        homeOfficeWorkLogs: [],
        timeOffWorkLogs: [],
      }),
      token: 'token',
    };
    const tree = shallow(<ListComponent {...props} />);

    expect(tree).toMatchSnapshot();
  });
});
