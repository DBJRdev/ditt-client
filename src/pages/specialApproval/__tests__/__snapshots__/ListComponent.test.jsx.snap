// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<Connect(LayoutComponent)
  loading={false}
  title="Special approvals"
>
  <div>
    You do not seem to have any pending special approvals.
  </div>
</Connect(LayoutComponent)>
`;
